tensorflow_probability/python/bijectors/gev_cdf.py

```diff
@@ -0,0 +1,218 @@
+          validate_args=validate_args,
+          forward_min_event_ndims=0,
+          parameters=parameters,
+          name=name)
+
+  @property
+  def loc(self):
+    """
+    The `loc` in `Y = exp(-t(X))`
+    where `t(x) = (1 + conc * (x - loc) / scale) ) ** (-1 / conc)`
+    when `conc != 0` and `exp(-(x - loc) / scale)` when `conc = 0`.`
+    """
+    return self._loc
+
+  @property
+  def scale(self):
+    """
+    The `scale` in `Y = exp(-t(X))`
+    where `t(x) = (1 + conc * (x - loc) / scale) ) ** (-1 / conc)`
+    when `conc != 0` and `exp(-(x - loc) / scale)` when `conc = 0`.`
+    """
+    return self._scale
+
+  @property
+  def concentration(self):
+    """
+    The `concentration` in `Y = exp(-t(X))`
+    where `t(x) = (1 + conc * (x - loc) / scale) ) ** (-1 / conc)`
+    when `conc != 0` and `exp(-(x - loc) / scale)` when `conc = 0`.`
+    """
+    return self._concentration
+
+  @classmethod
+  def _is_increasing(cls):
+    return True
+
+  def _forward(self, x):
+    with tf.control_dependencies(self._maybe_assert_valid_x(x)):
+      conc = tf.convert_to_tensor(self.concentration)
+
+      z = (x - self.loc) / self.scale
+
+      # for the numerical stability, we set a threshold `tol` that
+      # when `abs(conc) < tol`, gev converges to the Gumbel dist.
+      # check the following link for the estimation
+      # of the threshold (also for _inverse and _forward_log_det_jacobian)
+      # https://github.com/tensorflow/probability/pull/901#issuecomment-629857187
+      tol_target = 1e-5
+      rtol_target = 2. * tol_target / tf.math.abs(z)
+      atol_target = 2. * tol_target / (tf.math.exp(-z) * tf.math.abs(z))
+      tol = tf.math.minimum(tf.math.minimum(rtol_target, atol_target),
+                            tf.constant(1., dtype=z.dtype))
+      t = tf.where(tf.math.less_equal(tf.math.abs(conc), tol),
+                   tf.exp(-z),
+                   tf.exp(-tf.math.log1p(z * conc) / conc))
+      return tf.exp(-t)
+
+  def _inverse(self, y):
+    with tf.control_dependencies(self._maybe_assert_valid_y(y)):
+      t = -tf.math.log(y)
+
+      tol_target = 1e-5
+      rtol_target = 2. * tol_target / tf.math.abs(tf.math.log(t))
+      atol_target = 2. * tol_target / (tf.math.log(t) ** 2)
+      tol = tf.math.minimum(tf.math.minimum(rtol_target, atol_target),
+                            tf.constant(1., dtype=t.dtype))
+
+      conc = tf.convert_to_tensor(self.concentration)
+      z = tf.where(tf.math.less_equal(tf.math.abs(conc), tol),
+                   -tf.math.log(t),
+                   tf.math.expm1(-tf.math.log(t) * conc) / conc)
+
+      return self.loc + self.scale * z
+
+  def _inverse_log_det_jacobian(self, y):
+    with tf.control_dependencies(self._maybe_assert_valid_y(y)):
+      t = -tf.math.log(y)
+
+      conc = tf.convert_to_tensor(self.concentration)
+      log_dt = (-conc - 1.) * tf.math.log(t)
+
+      return tf.math.log(self.scale / y) + log_dt
+
+  def _forward_log_det_jacobian(self, x):
```

Comment:
We could omit this and rely on inverse_log_det_jacobian since the calculation looks more numerically stable. WDYT?

Reply:
Yes, that's a good idea.  `_inverse_log_det_jacobian` has a much more numerically stable calculation formula. Will use  `_inverse_log_det_jacobian`  instead.